crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -572,8 +544,137 @@ where
+
+            let (checkpoint, epoch) =
+                Self::index_checkpoint_and_epoch(&self.state, &checkpoint_data)
+                    .await
+                    .tap_err(|e| {
+                        error!(
+                            "Failed to index checkpoints {:?} with error: {}",
+                            checkpoint_data,
+                            e.to_string()
+                        );
+                    })?;
+            index_timer.stop_and_record();
+
+            // commit first epoch immediately, send other epochs to channel to be committed later.
+            if let Some(epoch) = epoch {
+                if epoch.last_epoch.is_none() {
+                    let epoch_db_guard = self.metrics.epoch_db_commit_latency.start_timer();
+                    info!("Persisting first epoch...");
+                    let mut persist_first_epoch_res = self.state.persist_epoch(&epoch).await;
+                    while persist_first_epoch_res.is_err() {
+                        warn!("Failed to persist first epoch, retrying...");
+                        persist_first_epoch_res = self.state.persist_epoch(&epoch).await;
+                    }
+                    epoch_db_guard.stop_and_record();
+                    self.metrics.total_epoch_committed.inc();
+                    info!("Persisted first epoch");
+                } else {
+                    let epoch_sender_guard = self.epoch_sender.lock().await;
+                    // NOTE: when the channel is full, epoch_sender_guard will wait until the channel has space.
+                    epoch_sender_guard.send(epoch).await.map_err(|e| {
+                        error!(
+                            "Failed to send indexed epoch to epoch commit handler with error {}",
+                            e.to_string()
+                        );
+                        IndexerError::MpscChannelError(e.to_string())
+                    })?;
+                    drop(epoch_sender_guard);
+                }
+            }
+
+            let checkpoint_sender_guard = self.checkpoint_sender.lock().await;
+            // NOTE: when the channel is full, checkpoint_sender_guard will wait until the channel has space.
+            // Checkpoints are sent sequentially to stick to the order of checkpoint sequence numbers.
+            checkpoint_sender_guard
+                .send(checkpoint)
+                .await
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "checkpoint channel send should not fail, but got error: {:?}",
+                        e
+                    )
+                });
+            drop(checkpoint_sender_guard);
+
+            // NOTE(gegaowp): today ws processing actually will block next checkpoint download,
+            // we can pipeline this as well in the future if needed
+            //
+            // Note: No need to send subscriptions when catching up.
```

Comment:
thought it twice, this after the split of reader & writer, the pub-sub will not work b/c only the writer has the data but only the readers are there to serve.
let's add a TODO here to resolve that.

Reply:
I thought about it, it's not an easy task to solve.
One idea is we support subscription in writer, and this type of subscription is going to very very simple, no fancy filters at all. Wdyt?